rmw_fastrtps_shared_cpp/src/rmw_get_topic_info.cpp

```diff
@@ -0,0 +1,306 @@
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+  // set topic type
+  ret = rmw_topic_info_set_topic_type(topic_info, topic_data.topic_type.c_str(), allocator);
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+  // Check if this participant is the same as the node that is passed
+  if (topic_data.participant_guid == participant_guid) {
+    ret = rmw_topic_info_set_node_name(topic_info, node_name, allocator);
+    if (ret != RMW_RET_OK) {
+      return ret;
+    }
+    ret = rmw_topic_info_set_node_namespace(topic_info, node_namespace, allocator);
+    if (ret != RMW_RET_OK) {
+      return ret;
+    }
+    return ret;
+  }
+  // This means that this discovered participant is not associated with the passed node
+  // and hence we must find its name and namespace from the discovered_names and
+  // discovered_namespace maps
+  // set node name
+  const auto & d_name_it = slave_target->discovered_names.find(topic_data.participant_guid);
+  if (d_name_it != slave_target->discovered_names.end()) {
+    ret = rmw_topic_info_set_node_name(topic_info, d_name_it->second.c_str(), allocator);
+  } else {
+    ret = rmw_topic_info_set_node_name(topic_info, "_NODE_NAME_UNKNOWN_", allocator);
+  }
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+  // set node namespace
+  const auto & d_namespace_it =
+    slave_target->discovered_namespaces.find(topic_data.participant_guid);
+  if (d_namespace_it != slave_target->discovered_namespaces.end()) {
+    ret = rmw_topic_info_set_node_namespace(topic_info, d_namespace_it->second.c_str(), allocator);
+  } else {
+    ret = rmw_topic_info_set_node_namespace(topic_info, "_NODE_NAMESPACE_UNKNOWN_", allocator);
+  }
+  return ret;
+}
+
+
+rmw_ret_t
+_get_info_by_topic(
+  const char * identifier,
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * topic_name,
+  bool no_mangle,
+  bool is_publisher,
+  rmw_topic_info_array_t * participants_info)
+{
+  rmw_ret_t ret = _validate_params(
+    identifier,
+    node,
+    allocator,
+    topic_name,
+    participants_info);
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+
+  const auto & topic_fqdns = _get_topic_fqdns(topic_name, no_mangle);
+  auto impl = static_cast<CustomParticipantInfo *>(node->data);
+  // The GUID of the participant associated with this node
+  const auto & participant_guid = impl->participant->getGuid();
+  const auto & node_name = node->name;
+  const auto & node_namespace = node->namespace_;
+  ::ParticipantListener * slave_target = impl->listener;
+  auto & topic_cache =
+    is_publisher ? slave_target->writer_topic_cache : slave_target->reader_topic_cache;
+  {
+    std::lock_guard<std::mutex> guard(topic_cache.getMutex());
+    const auto & topic_name_to_data = topic_cache().getTopicNameToTopicData();
+    std::vector<rmw_topic_info_t> topic_info_vector;
+    for (const auto & topic_name : topic_fqdns) {
+      const auto & it = topic_name_to_data.find(topic_name);
```

Comment:
@mm318 `it` is an rvalue, do not use an lvalue reference to hold it.

Reply:
Done.